### PR TITLE
[TT-7572] OAS authentication plugin

### PR DESCRIPTION
**Pull request by jeffy-mathew**
Related Issue - https://tyktech.atlassian.net/browse/TT-7572

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4667/merge
Commit: 0fb3932b6c60e936ffdc5f698c979d73c3ee469d 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3958316319)

**Review by titpetric — APPROVED** [review]

File: apidef/oas/middleware.go [hunk 1]
```diff
@@ -720,3 +737,34 @@ func (et *EnforceTimeout) ExtractTo(meta *apidef.HardTimeoutMeta) {
...
+	// Path is the path to shared object file in case of gopluign mode or path to js code in case of otto auth plugin.
+	Path string `bson:"path,omitempty" json:"path,omitempty"`
+	// RawBodyOnly if set to true, do not fill body in request or response object.
+	RawBodyOnly bool `bson:"rawBodyOnly,omitempty" json:"rawBodyOnly,omitempty"`
+}
+
+// AuthenticationPlugin holds the configuration for custom authentication plugin.
+type AuthenticationPlugin struct {
+	// Enabled enables custom authentication plugin.
+	Enabled bool `bson:"enabled,omitempty" json:"enabled,omitempty"` // required.
```
Comment (furkansenharputlu) [inline apidef/oas/middleware.go:753]: `enabled` should be omitted

File: apidef/oas/middleware.go [hunk 2]
```diff
@@ -720,3 +737,34 @@ func (et *EnforceTimeout) ExtractTo(meta *apidef.HardTimeoutMeta) {
 	meta.Disabled = !et.Enabled
 	meta.TimeOut = et.Value
 }
+
+type CustomPluginMiddleware struct {
+	// FunctionName is the name of authentication method.
+	FunctionName string `bson:"functionName,omitempty" json:"functionName,omitempty"`
```
Comment (furkansenharputlu) [inline apidef/oas/middleware.go:743]: Is there a default value for `functionName` and `path`? If not, why `omitempty`? Shouldn't they be required?

File: apidef/oas/middleware.go [hunk 3]
```diff
@@ -720,3 +737,34 @@ func (et *EnforceTimeout) ExtractTo(meta *apidef.HardTimeoutMeta) {
...
+	// RawBodyOnly if set to true, do not fill body in request or response object.
+	RawBodyOnly bool `bson:"rawBodyOnly,omitempty" json:"rawBodyOnly,omitempty"`
+}
+
+// AuthenticationPlugin holds the configuration for custom authentication plugin.
+type AuthenticationPlugin struct {
+	// Enabled enables custom authentication plugin.
+	Enabled bool `bson:"enabled,omitempty" json:"enabled,omitempty"` // required.
+	// CustomPluginMiddleware holds configuration for custom middleware.
+	CustomPluginMiddleware `bson:",inline" json:",inline"`
```
Comment (furkansenharputlu) [inline apidef/oas/middleware.go:755]: Why is this inlined? If it is used multiple times, why don't it include `enabled inside` ?

Comment (jeffy-mathew) [reply apidef/oas/middleware.go:755]: We can use `CustomPluginMiddleware` for other middlewares

File: apidef/oas/middleware.go (hunk 2 above)
Comment (jeffy-mathew) [reply apidef/oas/middleware.go:743]: yes, good catch. 
the function name and path - both are required.

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4667/merge
Commit: cfdf6317474d92ea81a3b2f22128cd46e11b4530 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3983646452)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4667/merge
Commit: 66d1696d7bebbbb5c4665a4af0704a6f6fd24074 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3983652526)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4667/merge
Commit: 341059c7941be83e4df9d06142270cadc11db638 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3986759424)

**Review by furkansenharputlu — APPROVED** [review]

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4667/merge
Commit: 1f9ceba7b99acc3b573cecf7979dbe7be41fef8b 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3986859469)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4667/merge
Commit: b2f38cece68b7bf5b492a52a53f250be972a51f0 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3986881998)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4667/merge
Commit: c8ab83388f3142587aa5088d587b18454c2db990 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3986926796)

**Pull request merged**